mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -3224,54 +3343,72 @@ static enum MAPISTATUS emsmdbp_object_sharing_metadata_property(struct emsmdbp_c
 	xml = talloc_asprintf_append(xml, "</Initiator>");
 	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
 
-	/*** Invitation ***/
-	xml = talloc_asprintf_append(xml, "<Invitation>");
-	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
-
-	/* This seems to be included when the calendar is not a default one */
-	xml = talloc_asprintf_append(xml, "<Title>%s</Title>", (char *)data_pointers[4]); /* PidLidSharingRemoteName */
-	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	/* Get recipients */
+	retval = emsmdbp_object_sharing_metadata_recipients_attr(emsmdbp_ctx, sharing_object, &recipients_attr, local_mem_ctx);
+	OPENCHANGE_RETVAL_IF(retval != MAPI_E_SUCCESS, retval, local_mem_ctx);
 
-	/*** Provider ***/
-	xml = talloc_asprintf_append(xml, "<Providers><Provider Type=\"ms-exchange-internal\" ");
-	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	/* See [MS-OXSHARE] Section 2.2.2.5 */
+	if (sharing_flavour == 0x23310) {
+		/*** Accept Of Request */
+		retval = emsmdbp_object_sharing_metadata_providers_node(emsmdbp_ctx, sharing_object,
+									retvals, data_pointers, true,
+									recipients_attr,
+									&providers_node, local_mem_ctx);
+		OPENCHANGE_RETVAL_IF(retval != MAPI_E_SUCCESS, retval, local_mem_ctx);
 
-	/* Get recipients */
-	contextID = emsmdbp_get_contextID(sharing_object);
-	mapistore_message_get_message_data(emsmdbp_ctx->mstore_ctx, contextID, sharing_object->backend_object, local_mem_ctx, &msg_data);
-	if (msg_data && msg_data->recipients_count > 0) {
-		xml = talloc_asprintf_append(xml, "TargetRecipients=\"");
+		xml = talloc_asprintf_append(xml, "<AcceptOfRequest>%s</AcceptOfRequest>", providers_node);
 		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
-
-		emsmdbp_fill_prop_index(&prop_index, msg_data->columns);
-		for (i = 0; i < msg_data->recipients_count; i++) {
-			if (prop_index.email_address != (uint32_t) -1) {
-				xml = talloc_asprintf_append(xml, "%s", (char *)msg_data->recipients[i].data[prop_index.email_address]);
+	} else if (sharing_flavour == 0x25100) {
+		/*** Deny of Request ***/
+		retval = emsmdbp_object_sharing_metadata_providers_node(emsmdbp_ctx, sharing_object,
+									retvals, data_pointers, false,
+									recipients_attr,
+									&providers_node, local_mem_ctx);
+		OPENCHANGE_RETVAL_IF(retval != MAPI_E_SUCCESS, retval, local_mem_ctx);
+
+		xml = talloc_asprintf_append(xml, "<DenyOfRequest>%s</DenyOfRequest>", providers_node);
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	} else {
+		if ((sharing_flavour & 0x20400) == 0x20400) {
```

Comment:
I'm afraid is not. Example, a sharing invitation for a special folder (0x20310):
```python
> hex(0x20310 & 0x20400)
'0x20000'
```